ngraph/core/reference/src/runtime/reference/fft.cpp

```diff
@@ -0,0 +1,533 @@
+                        {
+                            axis += complex_data_rank;
+                        }
+                    }
+                    return result;
+                }
+
+                std::vector<int64_t> get_axes(const int64_t* axes_data,
+                                              const Shape& axes_data_shape,
+                                              int64_t complex_data_rank)
+                {
+                    auto axes = canonicalize_axes(axes_data, axes_data_shape, complex_data_rank);
+                    std::sort(axes.begin(), axes.end(), std::greater<int64_t>{});
+                    return axes;
+                }
+
+                void reverse_fft_axes(std::vector<int64_t>& axes, int64_t complex_data_rank)
+                {
+                    for (int64_t& axis : axes)
+                    {
+                        axis = complex_data_rank - 1 - axis;
+                    }
+                }
+
+                std::vector<int64_t> get_lengths(const std::vector<int64_t>& shape,
+                                                 const std::vector<int64_t>& axes)
+                {
+                    std::vector<int64_t> lengths;
+                    for (int64_t axis : axes)
+                    {
+                        lengths.push_back(shape[axis]);
+                    }
+                    return lengths;
+                }
+
+                std::vector<int64_t> get_outer_axes(const std::vector<int64_t>& inner_axes,
+                                                    int64_t complex_data_rank)
+                {
+                    int64_t num_of_inner_axes = static_cast<int64_t>(inner_axes.size());
+                    int64_t num_of_outer_axes = complex_data_rank - num_of_inner_axes;
+
+                    std::vector<int64_t> outer_axes(num_of_outer_axes);
+
+                    int64_t fft_axes_as_bitset = 0;
+                    for (int64_t axis : inner_axes)
+                    {
+                        assert(axis < 64);
+                        fft_axes_as_bitset |= static_cast<int64_t>(1) << axis;
+                    }
+
+                    for (int64_t j = 0, i = 0; i < complex_data_rank; ++i)
+                    {
+                        if ((fft_axes_as_bitset & (static_cast<int64_t>(1) << i)) == 0)
+                        {
+                            outer_axes[j] = i;
+                            ++j;
+                        }
+                    }
+
+                    return outer_axes;
+                }
+
+                inline bool is_power_of_two(int64_t x) { return (x != 0) && ((x & (x - 1)) == 0); }
+
+                int64_t compute_buffer_size(const std::vector<int64_t>& fft_lengths)
+                {
+                    int64_t buffer_size = 0;
+
+                    for (int64_t length : fft_lengths)
+                    {
+                        int64_t current_size = is_power_of_two(length) ? (2 * length) : length;
```

Comment:
Why do we need a buffer twice as large for a power of two?

Reply:
Because we use Cooley-Tukey radix-2 FFT algorithm, when the length of the axis is a power of 2, and this algorithm requires a buffer twice as large. See https://en.wikipedia.org/wiki/Cooley%E2%80%93Tukey_FFT_algorithm. Next, we use this algorithm, because TensorFlow uses it, when the length of the axis is a power of 2.